Migrate jest tests to use Node.js 20 and make components compatible with upgraded packages

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #1098

To make this repo use Node.js 20, this PR migrates jest tests along with two rule fixes for eslint:

- Adjust the settings so that jest can run:
   - Fix that `is-plain-obj@4.1.0` doesn't provide the CommonJS build.
   - Force `uuid` to resolve with the CommonJS entry point as jest doesn't support `package.json.exports`.
   - Fix `@woocommerce/components` still using incompatible `@woocommerce/currency`.
   - Ignore that `qrcode.react` module is incorrectly listed in dev dependencies of `@automattic/components`.
   - Fix the React versioning conflicts between `@wordpress/*` and `@woocommerce/*`.
- Update timer mocks to deprecate the `timers` option.
- Migrate test cases to `@testing-library/user-event` v14.
   - Invoke `userEvent.setup()` in the test itself and before the component is rendered.
   - Adding `async` and `await` as all APIs are asynchronous.
   - No need to be wrapped in `act`.
   - Replace the simulation of pressing buttons with the `keyboard` API.
   - `clear()` is changed to only support editable elements only.
- Prevent form submission during testing.
- The ARIA role of `<img>` is considered as `presentation` if its `alt` is an empty string.
- Update snapshots to reflect the format changes.
- Fix `CampaignNameCell` tests by replacing `fireEvent` with `userEvent`.
- Resolve eslint errors for the `testing-library/no-wait-for-multiple-assertions` rule.
- Resolve eslint errors for the `testing-library/no-unnecessary-act` rule.
- Ignore the deprecation warnings of `defaultProps` from React by calling `console.error()`.
- Remove the `structuredClone` polyfill no longer needed.

In addition, it also needs a few adjustments to components or their tests to be compatible with upgraded `@wordpress/components` and `@woocommerce/components`:

- Prevent the `<Dashboard>` tests from calling `window.fetch`.
- Adjust `<Stepper>` related tests to wait for stepper content to be rendered.
- Make `<AppTooltip>` compatible with the `placement` prop introduced by WordPress 6.4 to replace `position`.
- Rewrite the replacer of `JSON.stringify` in `getReportKey` to ensure it returns the same key regardless of the query keys' order.
- Replace deprecated `event.keyCode` with `event.code` for `<VerificationCodeControl>`.
- Avoid accidentally using the event object to reset the asset group in `<CampaignAssetsForm>`.

#### 💡 The logical correlation between commits is low and highly independent, so it would be easier to review changes by looking at each commit individually. 🙏 

### Detailed test instructions:

1. Set Node.js to v20
2. Run `npm install` to see if it can finish without errors
3. Check if the following scripts work well
   - `npm run lint:js`
   - `npm run test:js`
4. Go to edit the phone number of contact information in the Setting page
   - In the verification code inputs, check if Arrow-Left, Arrow-Right and Backspace keyboards can move the input cursor as expected
5. Go to the Product Feed page
   - Hover to the edit icon on the Product Feed table
   - Check if the tooltip is shown and centered above the icon
   ![image](https://github.com/user-attachments/assets/1d49045a-f3aa-4c5b-89c9-224346ee467e)
6. View the e2e test result on GitHub Actions: https://github.com/woocommerce/google-listings-and-ads/actions/runs/10452211003/job/28940184227
7. Do some smoke testing

### Changelog entry

> Tweak - Make the Tooltip use the new placement prop when WordPress >= 6.4.
> Tweak - Replace deprecated event.keyCode with event.code for the verification code inputs in the contact information setting.
> Dev - Migrate jest tests to use Node.js 20.
> Dev - Avoid accidentally using the event object to reset the asset group values in the CampaignAssetsForm component.
> Dev - Rewrite the replacer of JSON.stringify in getReportKey to ensure it returns the same key regardless of the query keys' order.

eason9487 committed:
Jest: Fix that `is-plain-obj@4.1.0` doesn't provide the CommonJS build.

eason9487 committed:
Jest: Force 'uuid' to resolve with the CommonJS entry point as jest doesn't support `package.json.exports`.

eason9487 committed:
Jest: Fix `@woocommerce/components` still using incompatible `@woocommerce/currency`.

eason9487 committed:
Jest: Ignore that 'qrcode.react' module is incorrectly listed in dev dependencies of '@automattic/components'.

eason9487 committed:
Jest: Fix the React versioning conflicts between `@wordpress/*` and `@woocommerce/*`.

eason9487 committed:
Jest: Update timer mocks to deprecate the `timers` option.

eason9487 committed:
Jset: Migrate test cases to `@testing-library/user-event` v14.

Migration includes:
- Invoke `userEvent.setup()` in the test itself and before the component is rendered.
- Adding `async` and `await` as all APIs are asynchronous.
- No need to be wrapped in `act`.
- Replace the simulation of pressing buttons with the `keyboard` API.
- `clear()` is changed to only support editable elements only.

eason9487 committed:
Jest: Prevent form submission during testing.

eason9487 committed:
Jest: The ARIA role of `<img>` is considered as `presentation` if its `alt` is an empty string.

eason9487 committed:
Jest: Prevent the `<Dashboard>` tests from calling `window.fetch`.

eason9487 committed:
Jest: Adjust `<Stepper>` related tests to wait for stepper content to be rendered.

eason9487 committed:
Resolve eslint errors for the `testing-library/no-wait-for-multiple-assertions` rule.

eason9487 committed:
Resolve eslint errors for the `testing-library/no-unnecessary-act` rule.

eason9487 committed:
Jest: Update snapshots to reflect the format changes.

eason9487 committed:
Jest: Remove the `structuredClone` polyfill no longer needed.

eason9487 committed:
Make `<AppTooltip>` compatible with the `placement` prop introduced by WordPress 6.4 to replace `position`.

eason9487 committed:
Jest: Fix `CampaignNameCell` tests by replacing `fireEvent` with `userEvent`.

eason9487 committed:
Jest: Ignore the deprecation warnings of `defaultProps` from React by calling `console.error()`.

eason9487 committed:
Rewrite the replacer of `JSON.stringify` in `getReportKey` to ensure it returns the same key regardless of the query keys' order.

eason9487 committed:
Replace deprecated `event.keyCode` with `event.code` for `<VerificationCodeControl>`.

Ref: https://github.com/woocommerce/google-listings-and-ads/issues/1098

eason9487 committed:
Avoid accidentally using the event object to reset the asset group in `<CampaignAssetsForm>`.

puntope commented on js/src/tests/dependencies/woocommerce/settings.js:
```diff
@@ -1,3 +1,19 @@
+
+// Jest is currently running tests with npm packages bundled with WordPress 6.4.
+export function isWpVersion( version, operator ) {
+	if ( version === '6.4' && operator === '<' ) {
```
Not sure if this is correct, I expected some logic for comparing versions

eason9487 replied:
This mock function is used for jest testing only and therefore doesn't implement the actual comparison logic. Add a more detailed description in https://github.com/woocommerce/google-listings-and-ads/pull/2544/commits/c9a259fadc079c01303d7edc7c5535163783f6f5.

puntope replied:
Ah gotcha, I believe because the big amount of files I missed this was a mock for tests.

puntope commented on js/src/data/utils.js:
```diff
@@ -85,6 +85,18 @@ export function getReportQuery( category, type, query, dateReference ) {
 	return reportQuery;
 }
 
+function replacer( key, value ) {
```
Can we document what is the purpose of this function?

eason9487 replied:
Sure. Added in https://github.com/woocommerce/google-listings-and-ads/pull/2544/commits/69fc1b485b03de3c2e7cf5e7162013bd8f3fb588.

puntope commented on js/src/components/app-tooltip/index.js:
```diff
@@ -3,12 +3,34 @@
+
+// compatibility-code "WP < 6.4" -- `position` prop in Tooltip is deprecated since WP 6.4
+function mapPlacementToPosition( props ) {
+	if ( isWpVersion( '6.4', '<' ) ) {
```
This will be always false

eason9487 replied:
The result will depend on the actual version of installed WordPress, not the one mocked in `js/src/tests/dependencies/woocommerce/settings.js`.

eason9487 committed:
Add code document for the `replacer` function in js/src/data/utils.js.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2544#discussion_r1725024012

eason9487 committed:
Add a more detailed description for the mock `isWpVersion` function.

Ref: https://github.com/woocommerce/google-listings-and-ads/pull/2544#discussion_r1725007682

eason9487 commented:
Hi @puntope, thanks for the review. This PR is ready for a new round. Could you take a look?

puntope reviewed (approved):
✅  LGTM

Tested the next:

✅  Adjust the settings so that jest can run -- When these changes are not in place the tests return errors. 

- Fix that is-plain-obj@4.1.0 doesn't provide the CommonJS build.
- Force uuid to resolve with the CommonJS entry point as jest doesn't support package.json.exports.
- Fix @woocommerce/components still using incompatible @woocommerce/currency.
- Ignore that qrcode.react module is incorrectly listed in dev dependencies of @automattic/components.
- Fix the React versioning conflicts between @wordpress/* and @woocommerce/*.

✅  Update timer mocks to deprecate the timers option. 
✅  Migrate test cases to @testing-library/user-event v14.
✅  Smoke test Dashboard
✅  Smoke test <AppTooltip> 
✅  Smoke test reports 
✅  Smoke test phone verification code keyboard functionality.
